extensions/amp-a4a/0.1/amp-ad-template.js

```diff
@@ -0,0 +1,87 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AdResponseType,
+  ValidatorResult,
+} from './amp-ad-type-defs';
+import {AmpAdNetworkBase} from './amp-ad-network-base';
+import {NameFrameRenderer} from './name-frame-renderer';
+import {TemplateRenderer} from './template-renderer';
+import {TemplateValidator} from './template-validator';
+import {addParamToUrl} from '../../../src/url';
+import {dev} from '../../../src/log';
+import {startsWith} from '../../../src/string';
+
+// These have no side-effects, and so may be reused between all instances.
+const validator = new TemplateValidator();
+const nameFrameRenderer = new NameFrameRenderer();
+
+export const DATA_REQUEST_PARAM_PREFIX = 'requestParam';
+
+export class AmpAdTemplate extends AmpAdNetworkBase {
+  /**
+   * @param {!Element} element
+   */
+  constructor(element) {
+    super(element);
+
+    this.registerValidator(validator, AdResponseType.TEMPLATE);
+    this.registerRenderer(new TemplateRenderer(), ValidatorResult.AMP);
+    this.registerRenderer(nameFrameRenderer, ValidatorResult.NON_AMP);
+
+    /** @const {string} */
+    this.baseRequestUrl_ = element.getAttribute('src');
+    dev().assert(this.baseRequestUrl_, 'A src attribute must be specified.');
+
+    this.getContext().win = this.win;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.getContext().size = {
+      // TODO(levitzky) handle non-numeric values.
+      width: this.element.getAttribute('width'),
+      height: this.element.getAttribute('height'),
+      layout: this.element.getAttribute('layout'),
+    };
+  }
+
+  /** @override */
+  getRequestUrl() {
+    let url = this.baseRequestUrl_;
+
+    // We collect all fields in the dataset of the form
+    // 'data-request-var-<field_name>=<val>`, and append &<field_name>=<val> to
+    // the add request URL.
+    Object.keys(this.element.dataset).forEach(dataField => {
+      if (startsWith(dataField, DATA_REQUEST_PARAM_PREFIX)) {
+        const requestParamName = dataField.slice(
+            DATA_REQUEST_PARAM_PREFIX.length, dataField.length);
+        if (requestParamName) {
+          // Set the first character to lower case, as reading it in camelCase
+          // will automatically put it into upper case.
+          const finalParamName = requestParamName.charAt(0).toLowerCase() +
+              requestParamName.slice(1);
+          const finalParamValue = this.element.dataset[dataField];
+          url = addParamToUrl(url, finalParamName, finalParamValue);
+        }
+      }
+    });
+    this.getContext().adUrl = url;
```

Comment:
is it needed right now?

I want to make sure we only add necessary stuff into the `context` object. And I wish we can later document clearly what the context object contains.

Can you check all the other context data, and remove the ones are not used.

Reply:
Yes, this is used by friendly-frame-renderer.js on line 97, though I'm not certain how necessary that is.